### PR TITLE
WPS changes based on API reviews, PR reviews, and changes to LLC plans

### PR DESCRIPTION
**Pull request by KrzysztofCwalina**
- all method groups now have protocol methods (methods taking RequestOptions)
- changed semantics of string messages (they are serialized as plaintext by default)
- added overloads allowing to specify expiresAt, in addition to expiresAfter.
- small renames

### CODE REVIEW

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/GlobalSuppressions.cs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System.Diagnostics.CodeAnalysis;
+
+[assembly: SuppressMessage("Usage", "AZC0002:DO ensure all service methods, both asynchronous and synchronous, take an optional CancellationToken parameter called cancellationToken.", Justification = "CancellationToken can be passed through RequestOptions")]
```
Comment (ellismg) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/GlobalSuppressions.cs:6]: Opened https://github.com/Azure/azure-sdk-tools/issues/1715 so we don't lose track of fixing the analyzer here.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs [hunk 2]
```diff
@@ -26,22 +26,23 @@ internal partial class WebPubSubAuthenticationPolicy : HttpPipelineSynchronousPo
...
-        /// <param name="expireAfter"></param>
+        /// <param name="expiresAtUtc">DateTime.Kind must be DateTimeKind.Utc.</param>
         /// <param name="hmacSha512">SHA512 if true, otherwise SHA256</param>
         /// <returns></returns>
         public static string GenerateAccessToken(
             string audience,
             IEnumerable<Claim> claims,
             AzureKeyCredential key,
-            TimeSpan expireAfter = default,
+            DateTime expiresAtUtc = default,
```
Comment (ellismg) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs:36]: Why is this not `DateTimeOffset` ?

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/swagger/WebPubSub.json [hunk 3]
```diff
@@ -99,7 +99,7 @@
                     "webpubsub"
                 ],
                 "summary": "Check if the connection with the given connectionId exists.",
-                "operationId": "WebPubSubService_ConnectionExists",
+                "operationId": "WebPubSubService_ConnectionExistsImpl",
```
Comment (ellismg) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/swagger/WebPubSub.json:102]: We should be able to do this via a transform in the auto-rest MD (so you don't have to make this change in the swagger itself, so we can keep it in sync with the authoritative version.)

Maybe something like this in the autorest.md?


### Make WebPubSubPermission a regular enum
``` yaml
- from: swagger-document
  where: $.paths
  transform: >
    $["/api/hubs/{hub}/connections/{connectionId}"].head.operationId = "WebPubSubService_ConnectionExistsImpl";
    $["/api/hubs/{hub}/groups/{group}"].head.operationId = "WebPubSubService_GroupExistsImpl";
    $["/api/hubs/{hub}/users/{userId}"].head.operationId = "WebPubSubService_UserExistsImpl";
```

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs [hunk 4]
```diff
@@ -98,199 +96,180 @@ public WebPubSubServiceClient(string connectionString, string hub) : this(ParseC
...
 
         /// <summary>Broadcast message to all the connected client connections.</summary>
-        /// <param name="message"></param>
-        /// <param name="excluded">Excluded connection Ids.</param>
-        /// <param name="cancellationToken">The cancellation token to use.</param>
+        /// <param name="content"></param>
+        /// <param name="contentType">Defaults to ContentType.PlainText.</param>
         /// <returns>A <see cref="Response"/> if successful.</returns>
-        public virtual async Task<Response> SendToAllAsync(string message, IEnumerable<string> excluded = null, CancellationToken cancellationToken = default)
+        public virtual async Task<Response> SendToAllAsync(string content, ContentType contentType = default)
```
Comment (ellismg) [inline sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs:102]: I still miss having cancellation tokens on these high level methods, FWIW.

**Review by ellismg — COMMENTED** [review]
Some comments - don't really understand about the use of `DateTime` instead of `DateTimeOffset`.  Also, a small tip on a way to remove editing the swagger (and hopefully we can remove all of this ceremony when the code generator just generates the correct output shape for HEAD methods that are doing Exists checks.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs:36]: I feel like DT is better for representing UTC times, i.e. is has explicit support for UTC. Possibly I should change it to DTO to comply with guidelines.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/GlobalSuppressions.cs (hunk 1 above)
Comment (KrzysztofCwalina) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/GlobalSuppressions.cs:6]: Though, to be clear. I think this is triggered by my experiments in not having CT parameters in some helper overloads.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs (hunk 4 above)
Comment (KrzysztofCwalina) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubServiceClient_extensions.cs:102]: Let's chat more about it. What I am going after is a very common BCL pattern where only the most powerful overload takes CTs. See https://docs.microsoft.com/en-us/dotnet/api/system.io.stream.writeasync?view=net-5.0 as an example. 
I am not sure why we want to add CTs to all overloads.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/swagger/WebPubSub.json (hunk 3 above)
Comment (KrzysztofCwalina) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/swagger/WebPubSub.json:102]: Yeah, this is temporary. Having said that, this swagger is already forked (PR a week ago). It's forked to generate some methods as internal (and possibly some other small changes). We should work on getting all these customizations out and get back to the original swagger.

File: sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs (hunk 2 above)
Comment (KrzysztofCwalina) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs:36]: After a discussions, we decided to switch to DTO. But I will do it after the JWT implementation is changed. The current JWT library does not take in DTO.

Comment (ellismg) [reply sdk/webpubsub/Azure.Messaging.WebPubSub/src/WebPubSubAuthenticationPolicy_token.cs:36]: We talked about this offline and you are going to move to DTO with the next set of changes.

**Review by ellismg — APPROVED** [review]
LGTM,

**Pull request merged**